### PR TITLE
[llvm][OpenMP] Handle complex types in atomic read

### LINKED ISSUE #93441: [Flang][OpenMP] Compilation error when capture-statement described in atomic read construct contains a different type
```
Version of flang-new : 19.0.0(8930ba98e01bc66949e482b396f8389d64388359)/AArch64
```

When `capture-statement` described in `atomic read` construct contains a different `type`, the compilation terminates abnormally.

Compilation ends normally in the following cases:
- Do not specify the compilation option (`-fopenmp`)

The following are the test program, Flang-new, Gfortran and ifort compilation/execution result.

OMP_3_O_007_4_atm_type_006_re_22.f90:
```fortran
subroutine sub2()
  real(kind=4)    :: s_v_r2
  complex(kind=4) :: s_x_c2
  s_v_r2 = 0.0
  s_x_c2 = (1.0,2.0)
!$omp atomic read
   s_v_r2 = s_x_c2
!$omp end atomic
   write(6,*) "s_v_r2 = ", s_v_r2
end subroutine sub2

program main
  call sub2()
end program
```

```
$ flang-new -fopenmp OMP_3_O_007_4_atm_type_006_re_22.f90
flang-new: /lustre/home/ohno/LLVM_20240523/llvm-project/llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:5894: llvm::OpenMPIRBuilder::InsertPointTy llvm::OpenMPIRBuilder::createAtomicRead(const llvm::OpenMPIRBuilder::LocationDescription&, llvm::OpenMPIRBuilder::AtomicOpValue&, llvm::OpenMPIRBuilder::AtomicOpValue&, llvm::AtomicOrdering): Assertion `(XElemTy->isFloatingPointTy() || XElemTy->isIntegerTy() || XElemTy->isPointerTy()) && "OMP atomic read expected a scalar type"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /lustre/home/ohno/LLVM_20240523/release/bin/flang-new -fc1 -triple aarch64-unknown-linux-gnu -emit-obj -mrelocation-model pic -pic-level 2 -pic-is-pie -target-cpu generic -target-feature +outline-atomics -target-feature +v8a -target-feature +fp-armv8 -target-feature +neon -fopenmp -resource-dir /lustre/home/ohno/LLVM_20240523/release/bin/.. -mframe-pointer=non-leaf -o /tmp/OMP_3_O_007_4_atm_type_006_re_22-4aae60.o -x f95-cpp-input OMP_3_O_007_4_atm_type_006_re_22.f90
 #0 0x0000000004251280 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x4251280)
 #1 0x000000000424f188 llvm::sys::RunSignalHandlers() (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x424f188)
 #2 0x000000000424f30c SignalHandler(int) Signals.cpp:0:0
 #3 0x000040002da507a0 (linux-vdso.so.1+0x7a0)
 #4 0x000040002def6274 raise (/lib64/libc.so.6+0x36274)
 #5 0x000040002dee0a2c abort (/lib64/libc.so.6+0x20a2c)
 #6 0x000040002deefba0 __assert_fail_base (/lib64/libc.so.6+0x2fba0)
 #7 0x000040002deefc18 __assert_perror_fail (/lib64/libc.so.6+0x2fc18)
 #8 0x000000000728bf20 llvm::OpenMPIRBuilder::emitTaskwaitImpl(llvm::OpenMPIRBuilder::LocationDescription const&) (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x728bf20)
 #9 0x00000000059d86cc convertHostOrTargetOperation(mlir::Operation*, llvm::IRBuilderBase&, mlir::LLVM::ModuleTranslation&) OpenMPToLLVMIRTranslation.cpp:0:0
#10 0x00000000059d9a5c (anonymous namespace)::OpenMPDialectLLVMIRTranslationInterface::convertOperation(mlir::Operation*, llvm::IRBuilderBase&, mlir::LLVM::ModuleTranslation&) const OpenMPToLLVMIRTranslation.cpp:0:0
#11 0x0000000006047370 mlir::LLVM::ModuleTranslation::convertOperation(mlir::Operation&, llvm::IRBuilderBase&, bool) (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x6047370)
#12 0x00000000060509ec mlir::LLVM::ModuleTranslation::convertBlockImpl(mlir::Block&, bool, llvm::IRBuilderBase&, bool) (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x60509ec)
#13 0x00000000060516a0 mlir::LLVM::ModuleTranslation::convertOneFunction(mlir::LLVM::LLVMFuncOp) (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x60516a0)
#14 0x000000000605200c mlir::LLVM::ModuleTranslation::convertFunctions() (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x605200c)
#15 0x00000000060569d8 mlir::translateModuleToLLVMIR(mlir::Operation*, llvm::LLVMContext&, llvm::StringRef) (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x60569d8)
#16 0x00000000048c3098 Fortran::frontend::CodeGenAction::generateLLVMIR() (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x48c3098)
#17 0x00000000048c53c4 Fortran::frontend::CodeGenAction::executeAction() (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x48c53c4)
#18 0x0000000004285cdc Fortran::frontend::FrontendAction::execute() (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x4285cdc)
#19 0x000000000427ab04 Fortran::frontend::CompilerInstance::executeAction(Fortran::frontend::FrontendAction&) (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x427ab04)
#20 0x000000000428b298 Fortran::frontend::executeCompilerInvocation(Fortran::frontend::CompilerInstance*) (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x428b298)
#21 0x0000000003e886d0 fc1_main(llvm::ArrayRef<char const*>, char const*) (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x3e886d0)
#22 0x0000000003dfc138 main (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x3dfc138)
#23 0x000040002dee4384 __libc_start_main (/lib64/libc.so.6+0x24384)
#24 0x0000000003e87554 _start (/lustre/home/ohno/LLVM_20240523/release/bin/flang-new+0x3e87554)
flang-new: error: unable to execute command: Aborted (core dumped)
flang-new: error: flang frontend command failed due to signal (use -v to see invocation)
flang-new version 19.0.0git (https://github.com/llvm/llvm-project.git 8930ba98e01bc66949e482b396f8389d64388359)
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /lustre/home/ohno/LLVM_20240523/release/bin
Build config: +assertions
flang-new: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
flang-new: note: diagnostic msg: /tmp/OMP_3_O_007_4_atm_type_006_re_22-0ff303
flang-new: note: diagnostic msg: /tmp/OMP_3_O_007_4_atm_type_006_re_22-0ff303.sh
flang-new: note: diagnostic msg:

********************
$
```

```
$ cat /tmp/OMP_3_O_007_4_atm_type_006_re_22-0ff303
#line "./OMP_3_O_007_4_atm_type_006_re_22.f90" 1
      subroutine sub2()
      real(kind=4) :: s_v_r2
      complex(kind=4) :: s_x_c2
      s_v_r2 = 0.0
      s_x_c2 =(1.0,2.0)
!$omp atomic read
      s_v_r2 = s_x_c2
!$omp end atomic
      write(6,*) "s_v_r2 = ", s_v_r2
      end subroutine sub2

      program main
      call sub2()
      end program
```

```
$ cat /tmp/OMP_3_O_007_4_atm_type_006_re_22-0ff303.sh
# Crash reproducer for clang version 19.0.0git (https://github.com/llvm/llvm-project.git 8930ba98e01bc66949e482b396f8389d64388359)
# Driver args: "-fopenmp" "OMP_3_O_007_4_atm_type_006_re_22.f90"
# Original command:  "/lustre/home/ohno/LLVM_20240523/release/bin/flang-new" "-fc1" "-triple" "aarch64-unknown-linux-gnu" "-emit-obj" "-mrelocation-model" "pic" "-pic-level" "2" "-pic-is-pie" "-target-cpu" "generic" "-target-feature" "+outline-atomics" "-target-feature" "+v8a" "-target-feature" "+fp-armv8" "-target-feature" "+neon" "-fopenmp" "-resource-dir" "/lustre/home/ohno/LLVM_20240523/release/bin/.." "-mframe-pointer=non-leaf" "-o" "/tmp/OMP_3_O_007_4_atm_type_006_re_22-4aae60.o" "-x" "f95-cpp-input" "OMP_3_O_007_4_atm_type_006_re_22.f90"
 "/lustre/home/ohno/LLVM_20240523/release/bin/flang-new" "-fc1" "-triple" "aarch64-unknown-linux-gnu" "-emit-obj" "-mrelocation-model" "pic" "-pic-level" "2" "-pic-is-pie" "-target-cpu" "generic" "-target-feature" "+outline-atomics" "-target-feature" "+v8a" "-target-feature" "+fp-armv8" "-target-feature" "+neon" "-fopenmp" "-mframe-pointer=non-leaf" "-x" "f95-cpp-input" "OMP_3_O_007_4_atm_type_006_re_22-0ff303"
$
```

```
$ gfortran -fopenmp OMP_3_O_007_4_atm_type_006_re_22.f90
 s_v_r2 =    1.00000000
$
```

```
$ ifort -qopenmp -diag-disable=10448 OMP_3_O_007_4_atm_type_006_re_22.f90
 s_v_r2 =    1.000000
$
```

### PR DESCRIPTION
**Pull request by NimishMishra**
This patch adds functionality for atomically reading `llvm.struct` types.

Fixes: https://github.com/llvm/llvm-project/issues/93441

### CODE REVIEW

**Comment by tblah** [comment]
Thanks for your work on this.

One of our tests segfaults at runtime. Here is a minimal reproducer (that should loop forever)
```
program complex_read
  complex :: x = 0
  complex v
  !$omp parallel
  do
    !$omp atomic read
      v = x
    if (v /= x) exit
  end do
  !$omp end parallel
end program
```

I am testing on aarch64 (aws graviton 3) and using the atomic implementation from compiler-rt.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 1]
```diff
@@ -7832,6 +7832,18 @@ OpenMPIRBuilder::createAtomicRead(const LocationDescription &Loc,
...
+  } else if (XElemTy->isStructTy()) {
+    LoadInst *OldVal = Builder.CreateLoad(XElemTy, X.Var, "omp.atomic.read");
+    OldVal->setAtomic(AO);
+    const DataLayout &LoadDL = OldVal->getModule()->getDataLayout();
+    unsigned LoadSize =
+        LoadDL.getTypeStoreSize(OldVal->getPointerOperand()->getType());
+    OpenMPIRBuilder::AtomicInfo atomicInfo(
+        &Builder, XElemTy, LoadSize * 8, LoadSize * 8, OldVal->getAlign(),
+        OldVal->getAlign(), true /* UseLibcall */, X.Var);
+    auto AtomicLoadRes = atomicInfo.EmitAtomicLoadLibcall(AO);
```
Comment (Meinersbur) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:7844]: Libcalls should only be used if the device does not support `atomicrmw` of that size. E.g. when the struct is only 4 bytes long, emit `atomicrmw` instead of `call __atomic_load`.

I think this is part of the larger issue of handling atomics properly. For the immediate problem, this should work, but I suggest to add a TODO/FIXME comment here.

**Review by Meinersbur — COMMENTED** [review]
@tblah Is the segfault with this patch or without it?

**Comment by tblah** [comment]
> @tblah Is the segfault with this patch or without it?

The segfault is with this patch. Without this patch flang-new hits an assertion failure.

**Comment by NimishMishra** [comment]
Thanks @tblah for testing this PR. Without this patch, I believe the assertion would be related to emitting an atomic read on non-integer types. 

I'll investigate on this and get back.

**Comment by NimishMishra** [comment]
Hi @tblah,

    Thanks for the test. I could reproduce the issue with my patch.

    GDB shows the segfault is in atomic_load. I have a couple of questions regarding the test that can help in triaging this better:


I was not able to reproduce the segfault on the following test, which is strange, because I did not expect the segfault to depend on `do`

```
program complex_read
  complex :: x = 0
  complex v
  !$omp parallel
    !$omp atomic read
      v = x
  !$omp end parallel
end program
```

Moreover, the following test (where I remove the `if`), I observe a crash.

```
program complex_read
  complex :: x = 0
  complex v
  !$omp parallel
  do
    !$omp atomic read
      v = x
  end do
  !$omp end parallel
end program
```

Taking a step further, if I remove the `atomic read` (i.e. the following test), the crash still happens.

```
program complex_read
  complex :: x = 0
  complex v
  !$omp parallel
  do
      v = x
  end do
  !$omp end parallel
end program
```

**Comment by tblah** [comment]
Thanks for taking a look.

> GDB shows the segfault is in atomic_load. I have a couple of questions regarding the test that can help in triaging this better:

I have tried this using both gnu libatomic and compiler-rt so I expect the bug is in what we are passing to `atomic_load` rather than `atomic_load` itself.

I get the same results on your three examples:
1. Compiles and runs correctly
2. **compiler** crash in `llvm::CodeExtractor::findAllocas`
3. **compiler** crash in `llvm::CodeExtractor::findAllocas`

It isn't clear to me if the **runtime** crash in my test is the same bug or not. It might be. I have seen those compiler crashes before while making changes to `mlir/lib/Target/LLVMIR/OpenMPToLLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp` - if we generate malformed IR in the callbacks passed into `OpenMPIRBuilder` the code extractor can crash while trying to outline the openmp region.

I don't have a machine available to confirm, but I would guess (2) and (3) are reproducible on x86?

**Comment by tblah** [comment]
The LLVM IR for your final example right before the code extractor runs is as follows:
```
define void @_QQmain() #0 {
  %tid.addr = alloca i32, align 4
  %zero.addr = alloca i32, align 4
  %1 = alloca { float, float }, i64 1, align 8
  br label %entry

entry:                                            ; preds = %0
  %omp_global_thread_num = call i32 @__kmpc_global_thread_num(ptr @1)
  br label %omp.par.entry

omp.par.entry:                                    ; preds = %entry
  %tid.addr.local = alloca i32, align 4
  %tid = load i32, ptr %tid.addr.local, align 4
  %tid.addr.use = load i32, ptr %tid.addr, align 4
  %zero.addr.use = load i32, ptr %zero.addr, align 4
  br label %omp.reduction.init

omp.reduction.init:                               ; preds = %omp.par.entry
  br label %omp.par.region

omp.par.region:                                   ; preds = %omp.reduction.init
  br label %omp.par.region1

omp.par.region2:                                  ; preds = %omp.par.region2, %omp.par.region1
  %2 = load { float, float }, ptr @_QFEx, align 4
  store { float, float } %2, ptr %1, align 4
  br label %omp.par.region2

omp.par.region1:                                  ; preds = %omp.par.region
  br label %omp.par.region2

omp.region.cont:                                  ; No predecessors!
  br label %omp.par.pre_finalize

omp.par.pre_finalize:                             ; preds = %omp.region.cont
  br label %omp.par.outlined.exit

omp.par.outlined.exit:                            ; preds = %omp.par.pre_finalize
  br label %omp.par.exit.split

omp.par.exit.split:                               ; preds = %omp.par.outlined.exit
  unreachable
}
```

See the controlflow graph with `omp.par.region2` unconditionally branching to itself for the infinite `do` loop.

I think this is a different bug that the code extractor can't handle infinite loops in the control flow graph - it is looking for live out values from that CFG region but there's no way out of the region.

A much simpler example also crashes the compiler:
```
program infinite_loop
  integer :: i
  !$omp parallel
  do
    i = 0
  end do
  !$omp end parallel
end program
```

Surprisingly, equivalent C code does not crash the compiler.

In conclusion I think the compiler crash and the executable crash are different bugs. I will create a ticket for the compiler crash

**Comment by NimishMishra** [comment]
Hi @tblah,

   I have tested out this PR with working on https://github.com/llvm/llvm-project/issues/112908, and I do not see segfault (once we remove `do`). Is `atomic read` still segfaulting on aarch64?

**Review by tblah — APPROVED** [review]
Thanks for all the updates.

I still have a few failing tests on my end that I need to minimize and figure out if they are really bugs with the atomic implementation. But I think in the meantime it is better to merge this because it does work for most cases.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp (hunk 1 above)
Comment (NimishMishra) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:7860]: @Meinersbur I have added a FIXME for this point. Thanks. If everything else looks okay, can we merge this?
